### PR TITLE
Return CATID's for project system extenders

### PR DESCRIPTION
**Pull request by davkean**
Implement IExtenderCATIDProvider for C#/VB. F# is being tracked by https://github.com/dotnet/project-system/issues/3561.

Also filed https://github.com/dotnet/project-system/issues/3560 to implement support for ProjectConfigurationBrowseObject and implemented ConfigurationBrowseObject here https://devdiv.visualstudio.com/DevDiv/Default/_git/CPS/pullrequest/123753?_a=overview.

### CODE REVIEW

**Comment by davkean** [comment]
Note, behavior is identical to legacy with exception to; Project and ProjectItem objects. Legacy does not return a CATID for these but flavors do. Implemented them, so that flavors can just hang their extender properties directly on these.

**Comment by sharwell** [comment]
Is there any guidance for extension authors who have since updated their NuGet packages to provide XAML-based customizations for the Properties window? I'd like to avoid a situation where duplicate properties start showing - one through the extender APIs and the other through the XAML rules.

**Comment by davkean** [comment]
Extenders can opt in and out of objects that they extend via `CanExtend(object)`. Depending on the object you are extending, you should be able to walk the parent until you get to the owning project, at which point you can exclude CPS projects via https://github.com/Microsoft/VSProjectSystem/blob/master/doc/automation/finding_CPS_in_a_VS_project.md.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Automation/AbstractExtenderCATIDProvider.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                    return GetExtenderCATID(ExtendeeObject.FolderBrowseObject);
+
+                default:
+                case ExtenderCATIDType.Unknown:                                                 // EnvDTE.ProjectItem.Properties (when not file, folder, reference)
+                    BCLDebug.Assert(extenderCATIDType == ExtenderCATIDType.Unknown, $"Unrecognized CATID type {extenderCATIDType}");
+                    return null;
+            }
+        }
+
+        protected abstract string GetExtenderCATID(ExtendeeObject extendee);
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Automation/AbstractExtenderCATIDProvider.cs]: Should this take an `IProjectTree` for the node for when the second issue above is fixed?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Automation/AbstractExtenderCATIDProvider.cs]: No, 'IProjectTree' is not a good representation of all the objects we're referring to, some, like a configuration, don't have a representation in the tree `at all`. The DTE objects themselves are actually a better representation at this point.

**Pull request merged**